Support TakeOrderedAndProject

revans2 opened the pull request:
This fixes #1575 and fixes #103

I did not test topN with nested types going along for the ride (not being sorted on, but in the table that is being sorted), so I didn't turn them on.  They should work, but it was not a requirement.

My initial performance tests showed a little bit of a performance gain over not supporting take ordered and project, but it was not a huge amount of data so it is hard to tell.

revans2 commented:
build

jlowe commented on docs/FAQ.md:
```diff
@@ -17,6 +17,33 @@ with `collect`, `show` or `write` a new `DataFrame` is constructed causing Spark
+
+* `WholeStageCodeGen` - The GPU plan typically does not do code generation, and does not support generating code for an
+entire stage in the plan. Code generation is typically used to reduce the cost of processing data
+one row at a time. The GPU plan processes the data in a columnar format, so the costs are different.
```
Would be good to elaborate here, e.g.: mentioning GPU data is processed in batches of rows, and batches often contain many thousands of rows.

jlowe commented on docs/FAQ.md:
```diff
@@ -17,6 +17,33 @@ with `collect`, `show` or `write` a new `DataFrame` is constructed causing Spark
+one row at a time. The GPU plan processes the data in a columnar format, so the costs are different.
+
+* ColumnarToRow and RowToColumnar Transitions - The CPU version of Spark plans typically process data in a row based format. The main exception to
+this is reading some kinds of columnar data, like parquet. When transitioning between the CPU and
+the GPU require transitions between row and columnar formatted data.
```
```suggestion
this is reading some kinds of columnar data, like Parquet. Transitioning between the CPU and
the GPU requires transitioning between row and columnar formatted data.
```

jlowe commented on docs/FAQ.md:
```diff
@@ -17,6 +17,33 @@ with `collect`, `show` or `write` a new `DataFrame` is constructed causing Spark
+this is reading some kinds of columnar data, like parquet. When transitioning between the CPU and
+the GPU require transitions between row and columnar formatted data.
+
+* `GpuCoalesceBatches` and `GpuShuffleCoalesce` - Processing data on the GPU often scales sublinerarly. That means doubling the data does takes less
```
```suggestion
* `GpuCoalesceBatches` and `GpuShuffleCoalesce` - Processing data on the GPU scales sublinearly. That means doubling the data often takes less
```

sameerz replied:
Would it be helpful to add a sentence to indicate why processing on the GPU scales sublinearly?

jlowe commented on docs/FAQ.md:
```diff
@@ -17,6 +17,33 @@ with `collect`, `show` or `write` a new `DataFrame` is constructed causing Spark
+
+* `GpuCoalesceBatches` and `GpuShuffleCoalesce` - Processing data on the GPU often scales sublinerarly. That means doubling the data does takes less
+than half the time. Because of this we want to process larger batches of data when possible. These
+operators will try to build larger batches of data to process.
```
```suggestion
operators collect smaller batches of data into fewer, larger batches to process.
```

jlowe commented on docs/FAQ.md:
```diff
@@ -17,6 +17,33 @@ with `collect`, `show` or `write` a new `DataFrame` is constructed causing Spark
+
+* `SortMergeJoin` - The RAPIDS accelerator does not support sort merge joins yet. For now we
+translate sort merge joins into shuffled hash joins. Because of this there are times when
+Sorts may be removed or other sorts added to meet the ordering requirements of the query.
```
```suggestion
sorts may be removed or other sorts added to meet the ordering requirements of the query.
```

jlowe commented on integration_tests/src/main/python/data_gen.py:
```diff
@@ -221,13 +221,13 @@ def __init__(self, precision=None, scale=None, nullable=True, special_cases=[]):
         DECIMAL_MIN = Decimal('-' + ('9' * precision) + 'e' + str(-scale))
```
2021 copyrights

jlowe commented on integration_tests/src/main/python/sort_test.py:
```diff
@@ -34,6 +34,14 @@ def test_single_orderby(data_gen, order):
             lambda spark : unary_op_df(spark, data_gen).orderBy(order),
```
2021 copyrights

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/SpillableColumnarBatch.scala:
```diff
@@ -44,6 +44,8 @@ trait SpillableColumnarBatch extends AutoCloseable {
    *       with decompressing the data if necessary.
```
2021 copyrights

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/SpillableColumnarBatch.scala:
```diff
@@ -82,6 +85,11 @@ class SpillableColumnarBatchImpl (id: TempSpillBufferId,
   def spillId: TempSpillBufferId = id
 
+  override lazy val sizeInBytes: Long =
+    withResource(RapidsBufferCatalog.acquireBuffer(id)) { buff =>
```
Note it is not necessary to go through the buffer acquistion logic here.  The buffer is immutable in the catalog, and since we're not trying to access the actual data but only the metadata we don't need to acquire.  Might be worth adding a `RapidsBufferCatalog.getBufferSize(id)` method but not necessary as part of this PR.

revans2 replied:
That was a larger change than I really wanted to do, which is why I left it this way. Yes it is just meta data, but where the meta data is stored is in the RapidsBuffer. If I remember correctly that was because we initially thought about compressing the data when we spilled to host memory or disk. I didn't want to make a design change without verifying that it would be an OK thing to do.

jlowe replied:
Yes we originally thought we could mutate the buffer silently behind the scenes, but we've since moved to a setup where the `RapidsBufferId` describes an immutable piece of data.  That means compressing the data requires re-registering the buffer with a new buffer ID (and corresponding metadata).  We leaned towards this route because there were cases during shuffle where it would be problematic and surprising if a buffer contents mutated.

I'll file a followup to add a buffer size accessor to the catalog that doesn't require acquiring the buffer.

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/limit.scala:
```diff
@@ -19,15 +19,19 @@ package com.nvidia.spark.rapids
 import scala.collection.mutable.ArrayBuffer
```
2021 copyrights

sameerz commented on docs/FAQ.md:
```diff
@@ -17,6 +17,33 @@ with `collect`, `show` or `write` a new `DataFrame` is constructed causing Spark
+Typically, there is a one to one mapping between CPU stages in a plan and GPU stages.  There are a
+few places where this is not the case.
+
+* `WholeStageCodeGen` - The GPU plan typically does not do code generation, and does not support generating code for an
```
I'd like to keep the line lengths to 100 characters for ease of editing, where possible.

revans2 replied:
I can see us trying to limit ourselves to 100 characters most of the time, I have real problems from a change management perspective when we try to make all of the lines as close to 100 characters as possible, like most of this file is.  When you make a single small change to a line that forces it over 100 characters instantly you feel like you need to reformat the rest of the paragraph to make all of the lines as close to 100 characters as possible.  This means git sees a much bigger change and trying to merge conflicting changes becomes a real pain.  Especially because all of those changes have no impact to what the final rendered result looks like.

sameerz commented on docs/FAQ.md:
```diff
@@ -17,6 +17,33 @@ with `collect`, `show` or `write` a new `DataFrame` is constructed causing Spark
+
+* `TakeOrderedAndProject` - The `TakeOrderedAndProject` operator will take the top N entries in each task,
+shuffle the results to a single executor and then take the top N results from that. The GPU
+plan often has more metrics than the CPU versions do, and when we tried to combine all of these
+operations into a single stage the metrics where confusing to understand what was happening. Instead
+we split the single stage up into multiple smaller parts so the metrics are clearer.
```
```suggestion
plan often has more metrics than the CPU versions do, and when we tried to combine all of these
operations into a single stage the metrics were confusing to understand. Instead we split the single 
stage into multiple smaller parts so the metrics are clearer.
```

sameerz commented on docs/FAQ.md:
```diff
@@ -17,6 +17,33 @@ with `collect`, `show` or `write` a new `DataFrame` is constructed causing Spark
+entire stage in the plan. Code generation is typically used to reduce the cost of processing data
+one row at a time. The GPU plan processes the data in a columnar format, so the costs are different.
+
+* ColumnarToRow and RowToColumnar Transitions - The CPU version of Spark plans typically process data in a row based format. The main exception to
```
Should ColumnarToRow and RowToColumnar be backquoted, like, for example, `WholeStageCodeGen` is in the previous section?  

```suggestion
* `ColumnarToRow` and `RowToColumnar` transitions - The CPU version of Spark plans typically process data in a row based format. The main exception to
```

revans2 commented:
I addressed the review comments and updated the metrics a bit. I was seeing much linger times than I expected for `GpuTopN` total time and I found that I had `iter.next()` included in the total time, which resulted in also including the time it took for the upstream batch in some cases to be built.

revans2 commented:
build

jlowe commented:
build

revans2 commented:
@sameerz please take another look
